File: StubForFirebase/stub-publisher.js
const { initializeApp } = require('firebase/app');
const { getFirestore, collection, addDoc } = require('firebase/firestore');

// Firebase configuration (replace with your own or use environment variables)
const firebaseConfig = {
  apiKey: process.env.FIREBASE_API_KEY || "your-api-key",
  authDomain: process.env.FIREBASE_AUTH_DOMAIN || "your-project.firebaseapp.com",
  projectId: process.env.FIREBASE_PROJECT_ID || "your-project-id",
  storageBucket: process.env.FIREBASE_STORAGE_BUCKET || "your-project.appspot.com",
  messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID || "123456789",
  appId: process.env.FIREBASE_APP_ID || "1:123456789:web:abcdef"
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// Function to generate stub data based on the schema
function generateStubData() {
  const nodeId = generateMacAddress();
  const transmissionAttempts = Math.floor(Math.random() * 11); // 0-10
  const temperature = 20 + Math.random() * 10; // 20-30
  const humidity = 40 + Math.random() * 20; // 40-60
  const accelerometer = {
    x: (Math.random() - 0.5) * 2, // -1 to 1
    y: (Math.random() - 0.5) * 2,
    z: (Math.random() - 0.5) * 2
  };
  const gyroscope = {
    x: (Math.random() - 0.5) * 2,
    y: (Math.random() - 0.5) * 2,
    z: (Math.random() - 0.5) * 2
  };
  const barometricPressure = 1000 + Math.random() * 20; // 1000-1020
  const ledState = generateHexColor();

  return {
    node_id: nodeId,
    transmission_attempts: transmissionAttempts,
    environment: {
      temperature: parseFloat(temperature.toFixed(2)),
      humidity: parseFloat(humidity.toFixed(2)),
      accelerometer,
      gyroscope,
      barometric_pressure: parseFloat(barometricPressure.toFixed(2))
    },
    led_state: ledState
  };
}

// Helper function to generate a fake MAC address
function generateMacAddress() {
  const hexDigits = '0123456789ABCDEF';
  let mac = '';
  for (let i = 0; i < 6; i++) {
    mac += hexDigits.charAt(Math.floor(Math.random() * 16));
    mac += hexDigits.charAt(Math.floor(Math.random() * 16));
    if (i < 5) mac += ':';
  }
  return mac;
}

// Helper function to generate a random hex color
function generateHexColor() {
  const hexDigits = '0123456789ABCDEF';
  let color = '#';
  for (let i = 0; i < 6; i++) {
    color += hexDigits.charAt(Math.floor(Math.random() * 16));
  }
  return color;
}

// Function to publish data to Firebase
async function publishToFirebase(data, collectionName = 'iot-data') {
  try {
    const docRef = await addDoc(collection(db, collectionName), data);
    console.log('Document written with ID: ', docRef.id);
  } catch (e) {
    console.error('Error adding document: ', e);
  }
}

// Main function
async function main() {
  const collectionName = process.argv[2] || 'iot-data'; // Allow specifying collection via command line
  const stubData = generateStubData();
  console.log('Generated stub data:', JSON.stringify(stubData, null, 2));
  await publishToFirebase(stubData, collectionName);
}

main();

File: StubForFirebase/stub_publisher.py
import os
import json
import random
import firebase_admin
from firebase_admin import credentials, firestore

# Firebase configuration (use environment variables or service account key)
FIREBASE_CREDENTIALS_PATH = os.getenv('FIREBASE_CREDENTIALS_PATH', 'serviceAccountKey.json')
FIREBASE_COLLECTION = os.getenv('FIREBASE_COLLECTION', 'iot-data')

# Initialize Firebase
cred = credentials.Certificate(FIREBASE_CREDENTIALS_PATH)
firebase_admin.initialize_app(cred)
db = firestore.client()

def generate_mac_address():
    """Generate a fake MAC address."""
    return ':'.join(['{:02X}'.format(random.randint(0, 255)) for _ in range(6)])

def generate_hex_color():
    """Generate a random hex color."""
    return '#{:06X}'.format(random.randint(0, 0xFFFFFF))

def generate_stub_data():
    """Generate stub data based on the schema."""
    node_id = generate_mac_address()
    transmission_attempts = random.randint(0, 10)
    temperature = 20 + random.random() * 10  # 20-30
    humidity = 40 + random.random() * 20  # 40-60
    accelerometer = {
        'x': (random.random() - 0.5) * 2,
        'y': (random.random() - 0.5) * 2,
        'z': (random.random() - 0.5) * 2
    }
    gyroscope = {
        'x': (random.random() - 0.5) * 2,
        'y': (random.random() - 0.5) * 2,
        'z': (random.random() - 0.5) * 2
    }
    barometric_pressure = 1000 + random.random() * 20  # 1000-1020
    battery_voltage = round(3.0 + random.random() * 1.5, 2) if random.random() > 0.5 else None  # Optional, 3.0-4.5V or None
    led_state = generate_hex_color()

    data = {
        'node_id': node_id,
        'transmission_attempts': transmission_attempts,
        'environment': {
            'temperature': round(temperature, 2),
            'humidity': round(humidity, 2),
            'accelerometer': accelerometer,
            'gyroscope': gyroscope,
            'barometric_pressure': round(barometric_pressure, 2)
        },
        'led_state': led_state
    }
    if battery_voltage is not None:
        data['battery_voltage'] = battery_voltage
    return data

def publish_to_firebase(data, collection_name=FIREBASE_COLLECTION):
    """Publish data to Firebase Firestore."""
    try:
        doc_ref = db.collection(collection_name).add(data)
        print(f'Document written with ID: {doc_ref[1].id}')
    except Exception as e:
        print(f'Error adding document: {e}')

def main():
    collection_name = os.sys.argv[1] if len(os.sys.argv) > 1 else FIREBASE_COLLECTION
    stub_data = generate_stub_data()
    print('Generated stub data:', json.dumps(stub_data, indent=2))
    publish_to_firebase(stub_data, collection_name)

if __name__ == '__main__':
    main()